File: Vs3_181RDB178.py
import sys
import re

class automobilis:
    def __init__(self, marka='default', gads=0, parnesumkarba='default',registracijasNr='default', nobraukums=0):
        self._marka=marka
        self._gads=gads
        self._parnesumkarba=parnesumkarba
        self._registracijasNr=registracijasNr
        self._nobraukums=nobraukums
    def show(self):
        return '{:16}'.format(self._marka)+ " | "+ '{:8}'.format(self._gads)+ " | "+ '{:15}'.format(self._parnesumkarba) +" | "+ '{:18}'.format(self._registracijasNr)+ " | "+ '{:14}'.format(self._nobraukums)
    def saveErr(self):
        return self._marka+ " "+ self._gads +" "+ self._parnesumkarba +" "+ self._registracijasNr +" "+self._nobraukums
    def marka(self):
        return self._marka
    def gads(self):
        return self._gads
    def parnesumkarba(self):
        return self._parnesumkarba
    def registracijasNr(self):
        return self._registracijasNr
    def nobraukums(self):
        return self._nobraukums
    def dashRegistracijasNrCheck(self):
        garums=len(self._registracijasNr)
        if self._registracijasNr[2]=='-':
            regNrList=self._registracijasNr.split('-')
            if re.match(r"[A-Z][A-Z]$", regNrList[0]) and int(regNrList[1])>=1 and int(regNrList[1])<=9999:
                return 0
            else: 
                return 1
        else:
            return 1
    def spaceRegistracijasNrCheck(self):
        if " " in self._registracijasNr:
            return 1
        else:
            return 0
    def parnesumkarbaCheck(self):
        if re.match(r"[am]$", self._parnesumkarba):
            return 0
        else:
            return 1
    def nobraukumsCheck(self):
        try:
            int(self._nobraukums)
            return 0 
        except ValueError:
            return 1
    def spaceParnesumKarba(self):
        if ' ' in self._parnesumkarba:
            return 1
        else:
            return 0
    def spaceNobraukumsCheck(self):
        if ' ' in self._nobraukums:
            return 1
        else:
            return 0
    def gadsCheck(self):
        try:
            int(self._gads)
            if int(self._gads)>=1600 and int(self._gads)<=2020:
                return 0
        except ValueError:
            return 1
    def spaceGadsCheck(self):
        if ' ' in self._gads:
            return 1
        else:
            return 0
    def spaceMarka(self):
        x=len(self._marka)
        if self._marka[0]==' ' or self._marka[x-1]==' ':
            return 1
        else:
            return 0
def formateshana():
    print(f'{"Marka":<18s}{ "Gads":<10s}{"Parnesumkarba":<18s}{"RegistracijasNRr":<20s}{"Nobraukums":<14s}')
    print("-----------------------------------------------------------------------------")
    


if __name__ == '__main__':
    
    print("Jolanta Bērziņa 181RDB178  8.grupa")
    print (" ")

    while True:
        try:
           fails=input("Ievadiet faila nosaukumu: \n")
           INPUTFILE = fails
           dokuments=open(INPUTFILE,'r')
           break
        except IOError:
           print("Tada faila nav. Meginiet velreiz.\n")
           continue
    open('err.txt', 'w' ).close()
    automasinas=list()
    derigiIeraksti=list()

    for line in dokuments.readlines():
        newlist=line.split(';')

        if len(newlist)!=5:
            errlist=newlist
            with open("err.txt", "a" ) as file:
                file.write(" ".join(errlist)+'\n')
            file.close()
            continue
        marka,gads,parensumkarba,registracijasNr,nobraukums=newlist
        automasinas.append(automobilis(marka,gads,parensumkarba,registracijasNr,nobraukums))
    dokuments.close()
    formateshana()
    with open("err.txt", "a") as file:
        for n in automasinas:
            if n.spaceMarka()!=0:
                file.write(n.saveErr()+'\n')
            elif n.gadsCheck()!=0 or n.spaceGadsCheck()!=0:
                file.write(n.saveErr()+'\n')
            elif n.parnesumkarbaCheck()!=0 or n.spaceParnesumKarba()!=0 :
                file.write(n.saveErr()+'\n')
            elif n.spaceRegistracijasNrCheck()!=0 or n.dashRegistracijasNrCheck()!=0:
                file.write(n.saveErr()+'\n')
            elif n.nobraukumsCheck()!=0 or n.spaceNobraukumsCheck()!=0:
                file.write(n.saveErr()+'\n')
            else:
                derigiIeraksti.append(n)
                print(n.show())
    file.close()
    sys.stdout.flush()
    while True:
        print("Izvelaties vienu no meklesanas kriterijiem:\n")
        print(" Atrast automasinas pec markas - 1")
        print(" Atrast automasinas pec to parnesumkarbas veida - 2")
        print (" Atrast automasinas pec to izlaisanas gada - 3")
        print(" Partraukt programmas darbu - 4")

        try:
            choice=int(input("Meklet pec : \n"))
        except ValueError:
            print("Tika ievadita nepareiza vertiba. Meginiet velreiz. \n")
            continue

        if choice==1:
            markaUserInput=input("Ievadiet automasinas markas nosaukumu: \n")
            formateshana()
           
            autoskaits=0
            for n in derigiIeraksti:
                if markaUserInput==n.marka():
                    autoskaits=autoskaits+1
                    print(n.show())
            if autoskaits==0:
                print("Netika atrasta nevina sadas markas automasina. Meginiet velreiz.\n" )
                continue
            continueChoice=input("Vai velaties turpinat veikt meklesanu? [y/n]\n")
            if continueChoice=="y" or continueChoice=="Y":
                continue
            else: 
                break
        
        elif choice==2:
            parnesumkarbaUserInput=input("\n Ievadiet parnesumkarbas tipu pec kura meklet [a/m]: \n")
            if re.match(r"[am]$", parnesumkarbaUserInput):
                formateshana()
                autoskaits=0
                for n in derigiIeraksti:
                    if parnesumkarbaUserInput==n.parnesumkarba():
                        autoskaits=autoskaits+1
                        print(n.show())
                if autoskaits==0:
                    print("Netika atrasta nevina sada automasina. Meginiet velreiz.\n" )
                    continue
                continueChoice=input("Vai velaties turpinat veikt meklesanu? [y/n]\n")
                if continueChoice=="y" or continueChoice=="Y":
                    continue
                else:
                    break
            else:  
                print("\n Nepareizi ievadits parnesumkarbas tips. Meginiet velreiz.\n")
                continue
        
        elif choice==3:
            
            try:
                gadsUserInput=int(input("\n Ievadiet automasinas izlaisanas gadu: \n" ))
            except ValueError:
                print("Tika ievadits nepareizs izlaiduma gads. Meginiet velreiz.\n")
                continue

            autoskaits=0
            formateshana()
            for n in   derigiIeraksti:
                if int(gadsUserInput)<=int(n.gads()):
                        autoskaits=autoskaits+1
                        print(n.show())
            if autoskaits==0:
                print("Netika atrasta nevina sada automasina. Meginiet velreiz.\n" )
                continue
            continueChoice=input("Vai velaties turpinat veikt meklesanu? [y/n]")
            if continueChoice=="y" or continueChoice=="Y":
                continue
            else:
                break
            
        elif choice==4:
            print("Programma tiek partraukta.")
            break
        else:
            print("\n Tika ievadita nepareiza vertiba. Meginiet velreiz.\n")
    
    

File: Vs3_181RDB178.cpp

#include <stdio.h>
#include <iostream>
#include <fstream>
#include <iomanip>
#include <string>
#include <vector>
#include<algorithm>
using namespace std;

bool checkIfString(string name) {
	bool upper = true;
	bool lower = true;
	for (int i = 0; i < name.size(); i++)
	{

		if ((int)(name[i]) < 65 || (int)(name[i]) > 90)  upper = false;
		if ((int)(name[i]) < 97 || (int)(name[i]) > 122) lower = false;
		if (upper == false && lower == true) {
			upper = true; continue;
		}
		if (upper == true && lower == false) {
			lower = true; continue;
		}
		if (upper == false || lower == false) {

			return false;
		}
		return true;
	}
}

bool checkIfNumber(const string& st)
{
	string::const_iterator it = st.begin();
	while (it != st.end() && isdigit(*it)) ++it;
	return !st.empty() && it == st.end();
}

class automobilis {
public:
	string marka;
	string gads;
	string parnesumkarba;
	string registracijasNr;
	string nobraukums;
	bool invalid = false;
	void setMarka(string fromDoc) {
		if (checkIfString(fromDoc) == false) {
			invalid = true;
		}
		marka = fromDoc;
	}

	void setGads(string fromDoc) {
		if (checkIfNumber(fromDoc) == false && fromDoc.length() != 4) {
			invalid = true;
		}
		gads = fromDoc;
	}

	void setParnesumkarba(string fromDoc) {
		if (checkIfString(fromDoc) == false && fromDoc.length() != 1 && (fromDoc != "a" || fromDoc != "m")) {
			invalid = true;
		}
		parnesumkarba = fromDoc;
	}

	void setRegistracijasNr(string fromDoc) {
		string burti = fromDoc.substr(0, 2);
		string cipari = fromDoc.substr(3);
		if (fromDoc.substr(2, 1) != "-" || checkIfNumber(cipari) == false || checkIfString(burti) == false) {
			invalid = true;
		}
		registracijasNr = fromDoc;
	}

	void setNobraukums(string fromDoc) {
		if (checkIfNumber(fromDoc) == false) {
			invalid = true;
		}
		nobraukums = fromDoc;
	}
	
	
	void print() {
		if (invalid == true)return;
		cout << setw(16)<<marka << "  ";
		cout << gads << " " << "\t";
		cout << parnesumkarba << "  " << "\t";
		cout << registracijasNr << "  " << "\t";
		cout << nobraukums << "\n";
	}

};

void formesana() {
	cout << " ----------------------------------------------------------- \n";
}
int userInputInteger() {
	bool ok = false; int x;
	cout << "\n";
	while (!ok)
	{
		cin >> x;
		if (!cin.fail() && (cin.peek() == EOF || cin.peek() == '\n') && (x==1 || x==2 || x==3))
		{
			ok = true; return x;
		}
		else
		{
			cin.clear();
			cin.ignore(256, '\n');
			cout << "\n Ievades kluda. Atkartoti ievadiet skaitli: " << endl;
		}
	}
	return -1;
}
string userInputString() {
	bool ok = false; string x;
	cout << "\n";
	while (!ok)
	{
		cin >> x;
		if (x == "y" || x == "Y" || x == "n" || x == "N")
		{
			ok = true; return x;
		}
		else
		{
			cin.clear();
			cin.ignore(256, '\n');
			cout << "\n Ievades kluda. Atkartoti ievadiet izveli [y/n] " << endl;
		}
	}return "";
}
string userInputGear() {
	bool ok = false; string x;
	cout << "\n";
	while (!ok)
	{
		cin >> x;
		if (x == "a" || x == "m")
		{
			ok = true; return x;
		}
		else
		{
			cin.clear();
			cin.ignore(256, '\n');
			cout << "\n Ievades kluda. Atkartoti ievadiet parnesumkarbas tipu [a/m] " << endl;
		}
	}
	return"";
}
string userInputYear() {
	bool ok = false; int x;
	string gads;
	cout << "\n";
	while (!ok)
	{
		cin >> x;
		gads = to_string(x);
		if (gads.length()==4 && !cin.fail() && (cin.peek() == EOF || cin.peek() == '\n') )
		{
			ok = true; return gads;
		}
		else
		{
			cin.clear();
			cin.ignore(256, '\n');
			cout << "\n Ievades kluda. Atkartoti ievadiet izlaidma gadu: " << endl;
		}
	}
	return "";
}

int main()
{
	cout << "Jolanta Berzina 8.grupa\n";
	cout << "Ievadiet faila nosaukumu.\n";
	string fileName;
	cin >> fileName;
	vector<string> rinda;
	ifstream in(fileName.c_str());
	if (!in)
	{
		cerr << "Fails netika atrasts : " << fileName << endl;
		return false;
	}
	string str;
	int carSkaits = 0;

	while (getline(in, str)) {
		if (str.size() > 0) {
			rinda.push_back(str);
			carSkaits++;
		}
	}
	in.close();

	automobilis* masivs = new automobilis[carSkaits];
	string word = "";
	int struktura = 0;
	
	for (int sk = 0; sk < carSkaits; sk++) {
		struktura = 0;
		for (auto x : rinda[sk]) {
			if (x == ';' || x == '\n') {
				struktura++;
				if (struktura == 1)masivs[sk].setMarka(word);
				if (struktura == 2)masivs[sk].setGads(word);
				if (struktura == 3)masivs[sk].setParnesumkarba(word);
				if (struktura == 4)masivs[sk].setRegistracijasNr(word);
				if (struktura == 5)masivs[sk].setNobraukums(word);
				word = "";
			}
			else {
				if (x == ' ')continue;
				word = word + x;
			}
		}
	}
	/*formesana();   izvada visus auto
	for (int i = 0; i < carSkaits; i++) { masivs[i].print(); }
	cout << "\n"; */
	string turpinat;
	do {
		cout << "Meklet automasinas pec markas - 1 \n";
		cout << "Meklet automasinas pec  parnesumkarbas tipa - 2 \n";
		cout << "Meklet automasinas, kas tika izlaistas noteikta gada - 3 \n";
		int izvele;
		izvele= userInputInteger();

		if (izvele == 1) {
			string choice1 = "y";
			while (choice1 == "y" || choice1 == "Y") /* mekle auto pec markas*/
			{
				cout << "\n Ievadiet automasinas marku:" << "\n";
				string searchMark;
				cin >> searchMark;
				formesana();
				bool atrasts = false;
				for (int i = 0; i < carSkaits; i++) {
					if (masivs[i].marka == searchMark) {
						atrasts = true;
						masivs[i].print();
					}
				}
				if (atrasts == true) {
					cout << "\n Vai velaties meklet citas markas automasinas?<Y/N>\n";
					choice1=userInputString();

				}
				else {
					cout << "\n Nav nevienas sadas markas automasinas.Vai velaties meklet citu?<Y/N>\n";
					choice1=userInputString();

				}
			}
		}
		else if (izvele == 2) {
			string choice2 = "y";
			while (choice2 == "y" || choice2 == "Y") /* mekle auto pec parnesumkarbas*/
			{
				cout << "\n Ievadiet automasinas parnesumkarbas tipu [a/m]:" << "\n";
				string searchGear;
				searchGear=userInputGear();
				bool atrasts = false;
				formesana();
				for (int i = 0; i < carSkaits; i++) {
					if (masivs[i].parnesumkarba == searchGear) {
						atrasts = true;
						masivs[i].print();
					}
				}
				if (atrasts == true) {
					cout << "\n Vai velaties meklet cita vieda parnesumkarbas automasinas?<Y/N>\n";
					choice2=userInputString();
				}
				else {
					cout << "\n Nav nevienas sadas automasinas.Vai velaties meklet citu?<Y/N>\n";
					choice2=userInputString();
				}
			}
		}
		else {
			string choice3 = "y";
			while (choice3 == "y" || choice3 == "Y") /* mekle auto pec gada*/
			{
				cout << "\n Ievadiet automasinas izlaiduma gadu:" << "\n";
				string searchYear;
				searchYear=userInputYear();
				formesana();
				bool atrasts = false;
				for (int i = 0; i < carSkaits; i++) {
					if (masivs[i].gads == searchYear) {
						atrasts = true;
						masivs[i].print();
					}
				}
				if (atrasts == true) {
					cout << "\n Vai velaties meklet cita gada automasinas?<Y/N>\n";
					choice3 = userInputString();

				}
				else {
					cout << "\n Nav nevienas noradita gada automasinas.Vai velaties meklet citu?<Y/N>\n";
					choice3=userInputString();

				}
			}
		}

		cout << "Vai velaties veikt citu meklesanu (y/n)?";
		turpinat= userInputString();

	} while (turpinat == "y" || turpinat == "Y");


	string directory;
	const size_t last_slash_idx = fileName.rfind('\\');
	if (string::npos != last_slash_idx)
	{
		directory = fileName.substr(0, last_slash_idx);
	}
	ofstream outfile;
	outfile.open(directory + "/err.txt");
	for (int i = 0; i < carSkaits; i++) {
		if (masivs[i].invalid == true)
			outfile << masivs[i].marka << " " << masivs[i].gads << " " << masivs[i].parnesumkarba << " " << masivs[i].registracijasNr << " " << masivs[i].nobraukums << "\n";
	}
	outfile.close();
	return 0;

}